File: apps/player/apps/native/Sources/Core/ApprovalLink.swift
import Foundation

// A link can select a request to review. It cannot change Server or approve it.
enum ApprovalLink {
    static func code(_ url: URL, server: ServerAddress) throws -> String {
        guard url.absoluteString.utf8.count <= 4096,
              let parts = URLComponents(url: url, resolvingAgainstBaseURL: false),
              parts.scheme == "kinosail", parts.host == "approve", parts.path.isEmpty,
              parts.user == nil, parts.password == nil, parts.port == nil, parts.fragment == nil,
              let query = parts.queryItems, query.count == 2,
              Set(query.map(\.name)) == ["server", "code"],
              let origin = query.first(where: { $0.name == "server" })?.value,
              let code = query.first(where: { $0.name == "code" })?.value else { throw ClientError.invalidInput("That TV sign-in link is invalid.") }
        guard try ServerAddress(origin) == server else {
            throw ClientError.invalidInput("That TV uses a different Server. Approve it in the Player connected to that Server.")
        }
        return try Input.code(code)
    }

    static func url(server: ServerAddress, code: String) throws -> URL {
        try server.mediaURL("/connect?code=\(Input.code(code))")
    }
}

File: apps/player/apps/native/src/core/approval-link.ts
import { normalizeServerURL } from './server-client';
import { approvalCode } from './device-approval';

export function approvalURL(server: string, code: string): string {
  return `${normalizeServerURL(server)}/connect?code=${approvalCode(code)}`;
}

// A scanned link can select a request, but never change the signed-in Server.
export function readApprovalLink(
  params: Record<string, string | string[] | undefined>,
  currentServer: string,
): string {
  if (Object.keys(params).some((key) => key !== 'server' && key !== 'code'))
    throw new Error('This sign-in link is invalid. Scan the TV again.');
  if (
    typeof params.server !== 'string' ||
    params.server.length > 2048 ||
    normalizeServerURL(params.server) !== normalizeServerURL(currentServer)
  )
    throw new Error(
      'This TV uses a different Server. Open the Player app connected to that Server, or approve in your browser.',
    );
  return approvalCode(params.code);
}
